### PR TITLE
#2353 Simulation can deal with an iteration step and a time step sepa…

### PR DESCRIPTION
**Pull request by barondim**
…rately

**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non regression tests were added (new model, new feature and bug fix)
- [x] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

**Comment by barondim** [comment]
- [ ] Awaiting validation and short informal tests before merging

**Comment by marcochiaramello** [comment]
Closes #2353

File: dynawo/sources/Common/Dictionaries/DYNError_en_GB.dic [hunk 1]
```diff
@@ -25,6 +25,7 @@ CriteriaNotChecked          =             one simulation's criteria is not respe
 OpenFileFailed              =             failed to open file %1%
 NoJobDefined                =             no job found in the .jobs file
 IncorrectDelay              =             inconsistent delay %1% at time %2% (max delay is %3%)
+IterationStepAndTimeStepAtTheSameTime =   can't run simulation with iteration step and time step at the same time
```
Comment (marcochiaramello) [inline dynawo/sources/Common/Dictionaries/DYNError_en_GB.dic:28]: message not very clear

File: dynawo/sources/API/JOB/JOBXmlHandler.cpp [hunk 2]
```diff
@@ -460,6 +460,8 @@ CurvesHandler::create(attributes_type const& attributes) {
   curves_ = shared_ptr<CurvesEntry>(new CurvesEntry());
   curves_->setInputFile(attributes["inputFile"]);
   curves_->setExportMode(attributes["exportMode"]);
+  curves_->setIterStep(attributes["iterStep"]);
```
Comment (rosiereflo) [inline dynawo/sources/API/JOB/JOBXmlHandler.cpp:463]: We should make sure that only one (or none) is defined here

File: dynawo/sources/API/JOB/JOBCurvesEntry.h [hunk 3]
```diff
@@ -66,10 +79,24 @@ class CurvesEntry {
    */
   void setExportMode(const std::string& exportMode);
 
+  /**
+   * @brief Iteration step attribute setter
+   * @param iterStep Integer step to dump curve values every N iterations
+   */
+  void setIterStep(boost::optional<int> iterStep);
```
Comment (rosiereflo) [inline dynawo/sources/API/JOB/JOBCurvesEntry.h:86]: Unit tests are missing

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/API/JOB/JOBCurvesEntry.cpp [hunk 4]
```diff
@@ -36,19 +36,39 @@ CurvesEntry::getExportMode() const {
   return exportMode_;
 }
 
+boost::optional<int>
+CurvesEntry::getIterStep() const {
```
Comment (gautierbureau) [inline dynawo/sources/API/JOB/JOBCurvesEntry.cpp:40]: Iter -> Iteration

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 5]
```diff
@@ -922,7 +922,14 @@ Simulation::simulate() {
         connectedComponents_ = data_->findConnectedComponents();
       }
     }
+
+    boost::optional<int> iterationStep = jobEntry_->getOutputsEntry()->getCurvesEntry()->getIterationStep();
+    boost::optional<double> timeStep = jobEntry_->getOutputsEntry()->getCurvesEntry()->getTimeStep();
+    if (iterationStep && timeStep) {
+      throw DYNError(DYN::Error::SIMULATION, IterationStepAndTimeStepBothDefined);
```
Comment (marcochiaramello) [inline dynawo/sources/Simulation/DYNSimulation.cpp:929]: I think this is a check on the consistency of what indicated in the xml, I am not sure it should be done in DYNSimulation but in the API part, if possible. @rosiereflo your opinion?

Comment (rosiereflo) [reply dynawo/sources/Simulation/DYNSimulation.cpp:929]: I totally agree, cf my comment on JobXmlHandler.cpp :)

**Comment by barondim** [comment]
Problème sur les NRTs. Ne pas merger. []

**Comment by marcochiaramello** [comment]
I do not remember if there is a doc on jobs file xml inputs, if yes, can we update it? 
Thanks!

**Review by marcochiaramello — APPROVED** [review]

**Comment by barondim** [comment]
> 

Good point. I have to update it. (3.3.1 section)

**Review by marcochiaramello — APPROVED** [review]

File: dynawo/sources/API/JOB/JOBXmlHandler.cpp [hunk 6]
```diff
@@ -462,6 +462,8 @@ CurvesHandler::create(attributes_type const& attributes) {
   curves_ = shared_ptr<CurvesEntry>(new CurvesEntry());
   curves_->setInputFile(attributes["inputFile"]);
   curves_->setExportMode(attributes["exportMode"]);
+  curves_->setIterationStep(attributes["iterationStep"]);
```
Comment (rosiereflo) [inline dynawo/sources/API/JOB/JOBXmlHandler.cpp:465]: The test that only one is defined should be done here

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/IEEE14.jobs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, you can obtain one at http://mozilla.org/MPL/2.0/.
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+    simulation tool for power systems.
+--><dyn:jobs xmlns:dyn="http://www.rte-france.com/dynawo">
+  <dyn:job name="IEEE14 - Iteration step">
+    <dyn:solver lib="dynawo_SolverIDA" parFile="solvers.par" parId="2"/>
+    <dyn:modeler compileDir="outputs1">
```
Comment (rosiereflo) [inline nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/IEEE14.jobs:16]: I would porpose outputsIterationStep and outputsTimeStep for the folder names

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/reference/outputs2/curves/curves.csv [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+0.000000;1.059970;1.045041;1.010001;1.017703;1.019553;1.070149;1.061530;1.089926;1.055982;1.051051;1.057013;1.055333;1.050525;1.035657;1.000000;232.329584;-16.691112;1.058824;1.436651;1.000000;40.126172;43.675435;1.048778;1.269364;1.000000;-0.074832;25.065946;1.011505;1.112250;1.000000;0.005609;12.802790;1.070149;1.461749;1.000000;-0.006407;17.569940;1.096374;1.194231;0.216990;0.126990;
```
Comment (rosiereflo) [inline nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/reference/outputs2/curves/curves.csv:2]: There is not 0.4s between the lines?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/reference/outputs2/curves/curves.csv (hunk 8 above)
Comment (barondim) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/reference/outputs2/curves/curves.csv:2]: The dumps defined in the task are not alone in the crv file. That's why we don't see the 0.4s minimum offset between each dumps.

Comment (rosiereflo) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/reference/outputs2/curves/curves.csv:2]: not sure I got your answer, the only reason I see to dump a line before the 0.4s would be an event, and that would be a lot of event?

Comment (barondim) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/reference/outputs2/curves/curves.csv]: In the Simulation::simulate() method, there is a lot of updateCurves calls besides the call with the timeStep. That's why there are that many dumps.

File: dynawo/sources/API/JOB/test/res/iterationStepAndTimeStepDefinedAtTheSameTime.jobs [hunk 9]
```diff
@@ -0,0 +1,30 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<!--
+    Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline dynawo/sources/API/JOB/test/res/iterationStepAndTimeStepDefinedAtTheSameTime.jobs:3]: Should be:
`Copyright (c) 2022, RTE (http://www.rte-france.com)
See AUTHORS.txt
All rights reserved.
This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, you can obtain one at http://mozilla.org/MPL/2.0/.
SPDX-License-Identifier: MPL-2.0

This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools
for power systems.`

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/IEEE14.crv [hunk 10]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_IterationAndTimeSteps/IEEE14.crv:3]: Same as above

**Review by rosiereflo — CHANGES_REQUESTED** [review]
Only copyrights to be updated and we should be ok

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**